### DISCUSSION TITLE
Trying to understand how gzip works in kube-rs

### DISCUSSION THREAD

**Discussion by goenning** [kube-rs/kube#1025] [Q&A] [answered]
I found the gzip feature and when searching the code it seems like it only adds the decompression layer. But who/where is the accept-encoding request header set?

my understanding is that api server will only compress if both the request header is set and the feature gate is enabled.

**Comment by clux** [comment, marked as answer]
As far as I've understood it, the [`DecompressionLayer`](https://docs.rs/tower-http/latest/tower_http/decompression/struct.DecompressionLayer.html) is in charge of setting the `Accept-Encoding` header to requests.

**Reply by clux** [reply]
Dug a bit deeper because I started doubting it (all of their doc test insert the header manually).

We construct a [`DecompressionLayer::new()`](https://github.com/tower-rs/tower-http/blob/66787bbd4221cf7a1720dd7b9f7c1f8c945452eb/tower-http/src/decompression/layer.rs#L27-L31) in [client/builder.rs](https://github.com/kube-rs/kube/blob/28213785f73e0d6607b308daac31f37e8d109b5e/kube-client/src/client/builder.rs#L116-L120) (with the `gzip = ["client", "tower-http/decompression-gzip"]` feature).

This `new` method defers to [`AcceptEncoding::default`](https://github.com/tower-rs/tower-http/blob/aeb6e91a609723bd2ac0d8e67fe94574e5f1eb35/tower-http/src/compression_utils.rs#L95-L103), which allows all of the compression algorithms. This is later [filtered down to just gzip based on compiled-in features](https://github.com/tower-rs/tower-http/blob/aeb6e91a609723bd2ac0d8e67fe94574e5f1eb35/tower-http/src/compression_utils.rs#L27-L38).

The decompression layer then ensures the [`ACCEPT_ENCODING` header has an entry](https://github.com/tower-rs/tower-http/blob/66787bbd4221cf7a1720dd7b9f7c1f8c945452eb/tower-http/src/decompression/service.rs#L100-L105), so it should do what it says on the tin.

**Reply by goenning** [reply]
Ah right, that makes sense, I haven't looked at that crate, I was just looking inside kube-rs thinking that I'd have to set it myself. Thanks for digging that out!
